airflow/io/path.py

```diff
@@ -0,0 +1,399 @@
+
+    def size(self) -> int:
+        """Size in bytes of the file at this path."""
+        return self.fs.size(self.path)
+
+    def _cp_file(self, dst: ObjectStoragePath, **kwargs):
+        """Copy a single file from this path to another location by streaming the data."""
+        # create the directory or bucket if required
+        if dst.key.endswith(self.sep) or not dst.key:
+            dst.mkdir(exist_ok=True, parents=True)
+            dst = dst / self.key
+        elif dst.is_dir():
+            dst = dst / self.key
+
+        # streaming copy
+        with self.open("rb") as f1, dst.open("wb") as f2:
+            # make use of system dependent buffer size
+            shutil.copyfileobj(f1, f2, **kwargs)
+
+    def copy(self, dst: str | ObjectStoragePath, recursive: bool = False, **kwargs) -> None:
+        """Copy file(s) from this path to another location.
+
+        For remote to remote copies, the key used for the destination will be the same as the source.
+        So that s3://src_bucket/foo/bar will be copied to gcs://dst_bucket/foo/bar and not
+        gcs://dst_bucket/bar.
+
+        :param dst: Destination path
+        :param recursive: If True, copy directories recursively.
+
+        kwargs: Additional keyword arguments to be passed to the underlying implementation.
+        """
+        if isinstance(dst, str):
+            dst = ObjectStoragePath(dst)
+
+        # same -> same
+        if self.samestore(dst):
+            self.fs.copy(self.path, dst.path, recursive=recursive, **kwargs)
+            return
+
+        # use optimized path for local -> remote or remote -> local
+        if self.protocol == "file":
+            dst.fs.put(self.path, dst.path, recursive=recursive, **kwargs)
+            return
+
+        if dst.protocol == "file":
+            self.fs.get(self.path, dst.path, recursive=recursive, **kwargs)
+            return
+
+        if not self.exists():
+            raise FileNotFoundError(f"{self} does not exist")
+
+        # remote dir -> remote dir
+        if self.is_dir():
+            if dst.is_file():
+                raise ValueError("Cannot copy directory to a file.")
+
+            dst.mkdir(exist_ok=True, parents=True)
+
+            out = self.fs.expand_path(self.path, recursive=True, **kwargs)
+
+            for path in out:
+                # this check prevents one extra call to is_dir() as
+                # glob returns self as well
+                if path == self.path:
+                    continue
+
+                src_obj = ObjectStoragePath(path, conn_id=self._accessor._store.conn_id)
+
+                # skip directories, empty directories will not be created
+                if src_obj.is_dir():
+                    continue
+
+                src_obj._cp_file(dst)
+
+            return
+
+        # remote file -> remote dir
+        if self.is_file():
+            self._cp_file(dst, **kwargs)
+            return
```

Comment:
NIT: Should we remove if/return? If is not needed here - I believe there is no other option than to have is_file() true here (not is_dir() and exists() means that it is a file). And it feels weird to have both  - if without else and explicit return as the last thing in if.

Reply:
Or maybe just move it to explicit `else` :)

Reply:
Good catch!